File: proj3/benchmark/benchmark.go
package main

import (
	"math/rand"
	"proj3-redesigned/jacobi"
	"time"
	"math"
	"flag"
	"fmt"
)

func randomMatrix(rows int, cols int) [][]float64 {
	rand.Seed(time.Now().Unix())
	mat := make([][]float64, rows)

	for col := range mat {
		mat[col] = make([]float64, cols)
	}

	for r:=0;r<rows;r++ {
		sum := 0.0
		for c:=0;c<cols;c++ {
			mat[r][c] = float64(rand.Intn(1000))
			sum += math.Abs(mat[r][c])
		}
		mat[r][r] = sum + float64(rand.Intn(10))
	}

	return mat
}

func randomArray(size int) []float64 {
	rand.Seed(time.Now().Unix())
	arr := make([]float64, size)

	for s:=0;s<size;s++ {
		arr[s] = float64(rand.Intn(1000))
	}

	return arr
}

func main() {

	size := flag.Int("size", 100, "Size of the matrix (NxN)")
	rowsPerTask := flag.Int("rowsPerTask", 2, "Number of rows per task for work stealing")
	numThreads := flag.Int("numThreads", 2, "Number of threads to use")
	desiredConvergence := flag.Float64("convergence", 0.000001, "Desired convergence threshold")
	maxIter := flag.Int("maxIter", 10000, "Maximum number of iterations")
	mode := flag.String("mode", "sequential", "Execution mode: sequential, bsp, or worksteal")

	flag.Parse()

	A := randomMatrix(*size, *size)
	B := randomArray(*size)

	fmt.Printf("Running with size=%d, rowsPerTask=%d, numThreads=%d, convergence=%.8f, maxIter=%d\n",
		*size, *rowsPerTask, *numThreads, *desiredConvergence, *maxIter)

	start := time.Now()

	switch *mode {
	case "sequential":
		jacobi.Sequential(A, B, *desiredConvergence, *maxIter)
	case "bsp":
		jacobi.BSP(*numThreads, *maxIter, A, B, *desiredConvergence)
	case "worksteal":
		jacobi.WorkSteal(*numThreads, *rowsPerTask, *maxIter, A, B, *desiredConvergence)
	default:
		jacobi.Sequential(A, B, *desiredConvergence, *maxIter)
	}

	elapsed := time.Since(start)
	fmt.Printf("%.6f\n", elapsed.Seconds())
}

File: proj3/benchmark/benchmark.py
import subprocess
import matplotlib.pyplot as plt
import numpy as np

rows_per_task_options = [1, 2, 3, 4]
num_threads = [2, 4, 6, 8, 12]
desired_convergence = "0.000001"
max_iter = "1000"
matrix_size = "1024"
num_runs = 5

def get_average_runtime(mode, threads=None, rows_per_task=None):
    total_time = 0.0
    for i in range(num_runs):
        cmd = [
            "go", "run", "benchmark.go",
            f"-mode={mode}",
            f"-convergence={desired_convergence}",
            f"-maxIter={max_iter}",
            f"-size={matrix_size}"
        ]

        if threads is not None:
            cmd.append(f"-numThreads={threads}")
        if rows_per_task is not None:
            cmd.append(f"-rowsPerTask={rows_per_task}")

        print(f"Run {i+1}/{num_runs}: {' '.join(cmd)}")
        result = subprocess.run(cmd, capture_output=True, text=True)

        try:
            output = result.stdout.strip().splitlines()
            for line in reversed(output):
                try:
                    run_time = float(line)
                    total_time += run_time
                    break
                except ValueError:
                    continue
        except Exception as e:
            print(f"Failed to parse output:\n{result.stdout}\nError: {e}")
            return None

    return total_time / num_runs

print("\n=== Sequential Runs ===")
seq_time = get_average_runtime("sequential")
print(f"Sequential average time: {seq_time:.4f} seconds\n")

bsp_speedups = []
for t in num_threads:
    print(f"\n=== BSP with {t} threads ===")
    avg_bsp_time = get_average_runtime("bsp", threads=t)
    speedup = seq_time / avg_bsp_time
    bsp_speedups.append(speedup)

worksteal_speedups = {rpt: [] for rpt in rows_per_task_options}
for rpt in rows_per_task_options:
    for t in num_threads:
        print(f"\n=== WorkSteal with {t} threads and {rpt} rows per task ===")
        avg_ws_time = get_average_runtime("worksteal", threads=t, rows_per_task=rpt)
        speedup = seq_time / avg_ws_time
        worksteal_speedups[rpt].append(speedup)

plt.figure(figsize=(10, 6))

plt.plot(num_threads, bsp_speedups, marker='o', label='BSP')

for rpt, speedups in worksteal_speedups.items():
    plt.plot(num_threads, speedups, marker='o', label=f'WorkSteal (rows per thread={rpt})')

plt.xlabel("Threads")
plt.ylabel("Speedup (Sequential Time / Parallel Time)")
plt.title("Speedup vs Threads")
plt.legend()
plt.grid(True)
plt.savefig("speedup_plot.png")